dotnet/src/SemanticKernel.Abstractions/AI/FunctionChoiceBehaviors/RequiredFunctionChoiceBehavior.cs

```diff
@@ -0,0 +1,118 @@
+using System.Linq;
+
+namespace Microsoft.SemanticKernel;
+
+/// <summary>
+/// Represents <see cref="FunctionChoiceBehavior"/> that provides either all of the <see cref="Kernel"/>'s plugins' functions to AI model to call or specific ones.
+/// This behavior forces the model to always call one or more functions. The model will then select which function(s) to call.
+/// </summary>
+internal sealed class RequiredFunctionChoiceBehavior : FunctionChoiceBehavior
+{
+    /// <summary>
+    /// List of the functions to provide to LLM.
+    /// </summary>
+    private readonly IEnumerable<KernelFunction>? _functions;
+
+    /// <summary>
+    /// Indicates whether the functions should be automatically invoked by AI connectors.
+    /// </summary>
+    private readonly bool _autoInvoke;
+
+    /// <summary>
+    /// Functions selector to customize functions selection logic.
+    /// </summary>
+    private readonly Func<FunctionChoiceBehaviorFunctionsSelectorContext, IReadOnlyList<KernelFunction>?>? _functionsSelector;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AutoFunctionChoiceBehavior"/> class.
+    /// </summary>
+    /// <param name="functions">
+    /// Functions to provide to AI model. If null, all <see cref="Kernel"/>'s plugins' functions are provided to the model.
+    /// If empty, no functions are provided to the model, which is equivalent to disabling function calling.
+    /// </param>
+    /// <param name="autoInvoke">
+    /// Indicates whether the functions should be automatically invoked by AI connectors.
+    /// </param>
+    /// <param name="functionsSelector">
+    /// The callback function allows customization of function selection.
+    /// It accepts functions, chat history, and an optional kernel, and returns a list of functions to be used by the AI model.
+    /// This should be supplied to prevent the AI model from repeatedly calling functions even when the prompt has already been answered.
+    /// For example, if the AI model is prompted to calculate the sum of two numbers, 2 and 3, and is provided with the 'Add' function,
+    /// the model will keep calling the 'Add' function even if the sum - 5 - is already calculated, until the 'Add' function is no longer provided to the model.
+    /// In this example, the function selector can analyze chat history and decide not to advertise the 'Add' function anymore.
+    /// </param>
+    public RequiredFunctionChoiceBehavior(IEnumerable<KernelFunction>? functions = null, bool autoInvoke = true, Func<FunctionChoiceBehaviorFunctionsSelectorContext, IReadOnlyList<KernelFunction>?>? functionsSelector = null)
+    {
+        this._functions = functions;
+        this._autoInvoke = autoInvoke;
+        this._functionsSelector = functionsSelector;
+    }
+
+    /// <inheritdoc />
+#pragma warning disable SKEXP0001 // Type is for evaluation purposes only and is subject to change or removal in future updates. Suppress this diagnostic to proceed.
```

Comment:
Should this whole class also be an experimental one.

```suggestion
#pragma warning disable SKEXP0001 // Type is for evaluation purposes only and is subject to change or removal in future updates. Suppress this diagnostic to proceed.
```

Reply:
There's not much value in doing so because the class is internal.